text/0000-pub-restricted.md

```diff
@@ -0,0 +1,924 @@
+could think of was C++ `friend` modifiers (but `friend` is far more
+ad-hoc and free-form than what is being proposed here).
+
+### Scala
+
+It has since been pointed out to me that Scala has scoped access
+modifiers `protected[Y]` and `private[Y]`, which specify that access
+is provided upto `Y` (where `Y` can be a package, class or singleton
+object).
+
+The feature proposed by this RFC appears to be similar in intent to
+Scala's scoped access modifiers.
+
+Having said that, I will admit that I am not clear on what
+distinction, if any, Scala draws between `protected[Y]` and
+`private[Y]` when `Y` is a package, which is the main analogy for our
+purposes, or if they just allow both forms as synonyms for
+convenience.
+
+(I can imagine a hypothetical distinction in Scala when `Y` is a
+class, but my skimming online has not provided insight as to what the
+actual distinction is.)
+
+Even if there is some distinction drawn between the two forms in
+Scala, I suspect Rust does not need an analogous distinction in it's
+`pub(restricted)`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Obviously,
+`pub(restriction) item` complicates the surface syntax of the language.
+
+ * However, my counter-argument to this drawback is that this feature
+   in fact *simplifies* the developer's mental model. It is easier to
+   directly encode the expected visibility of an item via
+   `pub(restriction)` than to figure out the right concoction via a
+   mix of nested `mod` and `pub use` statements. And likewise, it is
+   easier to read it too.
+
+Developers may misuse this form and make it hard to access the tasty
+innards of other modules.
+
+ * This is true, but I claim it is irrelevant.
+
+   The effect of this change is solely on the visibility of items
+   *within* a crate. No rules for inter-crate access change.
+
+   From the perspective of cross-crate development, this RFC changes
+   nothing, except that it may lead some crate authors to make some
+   things no longer universally `pub` that they were forced to make
+   visible before due to earlier limitations. I claim that in such
+   cases, those crate authors probably always intended for such items
+   to be non-`pub`, but language limitations were forcing their hand.
+
+   As for intra-crate access: My expectation is that an individual
+   crate will be made by a team of developers who can work out what
+   mutual visibility they want and how it should evolve over time.
+   This feature may affect their work flow to some degree, but they
+   can choose to either use it or not, based on their own internal
+   policies.
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Do not extend the language!
+
+ * Change privacy rules and make privacy analysis "smarter"
+   (e.g. global reachabiliy analysis)
+
+   The main problem with this approach is that we tried it, and it
+   did not work well: The implementation was buggy, and the user-visible
+   error messages were hard to understand.
```

Comment:
Note, that reachability analysis is already performed by the privacy pass. Its results are used by rustdoc, dead code lint, stability and link-time-reachability passes, but not by privacy itself.
It's also less buggy than it was.